tianshou/env/venvs.py

```diff
@@ -0,0 +1,319 @@
+        observations if id is ``None``, otherwise reset the specific
+        environments with the given id, either an int or a list.
+        """
+        self._assert_is_not_closed()
+        id = self._wrap_id(id)
+        if self.is_async:
+            self._assert_id(id)
+        obs = np.stack([self.workers[i].reset() for i in id])
+        return obs
+
+    def step(self,
+             action: Optional[np.ndarray],
+             id: Optional[Union[int, List[int]]] = None
+             ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+        """Run one timestep of all the environments’ dynamics if id is "None",
+        otherwise run one timestep for some environments with given id,  either
+        an int or a list. When the end of episode is reached, you are
+        responsible for calling reset(id) to reset this environment’s state.
+
+        Accept a batch of action and return a tuple (obs, rew, done, info).
+
+        :param numpy.ndarray action: a batch of action provided by the agent.
+
+        :return: A tuple including four items:
+
+            * ``obs`` a numpy.ndarray, the agent's observation of current \
+                environments
+            * ``rew`` a numpy.ndarray, the amount of rewards returned after \
+                previous actions
+            * ``done`` a numpy.ndarray, whether these episodes have ended, in \
+                which case further step() calls will return undefined results
+            * ``info`` a numpy.ndarray, contains auxiliary diagnostic \
+                information (helpful for debugging, and sometimes learning)
+
+        For the async simulation:
+
+        Provide the given action to the environments. The action sequence
+        should correspond to the ``id`` argument, and the ``id`` argument
+        should be a subset of the ``env_id`` in the last returned ``info``
+        (initially they are env_ids of all the environments). If action is
+        ``None``, fetch unfinished step() calls instead.
+        """
+        self._assert_is_not_closed()
+        id = self._wrap_id(id)
+        if not self.is_async:
+            assert len(action) == len(id)
+            for i, j in enumerate(id):
+                self.workers[j].send_action(action[i])
+            result = [self.workers[j].get_result() for j in id]
+        else:
+            if action is not None:
+                self._assert_id(id)
+                assert len(action) == len(id)
+                for i, (act, env_id) in enumerate(zip(action, id)):
+                    self.workers[env_id].send_action(act)
+                    self.waiting_conn.append(self.workers[env_id])
+                    self.waiting_id.append(env_id)
+                self.ready_id = [x for x in self.ready_id if x not in id]
+            result = []
+            while len(self.waiting_conn) > 0 and len(result) < self.wait_num:
+                ready_conns = self.worker_class.wait(self.waiting_conn)
```

Comment:
`waiting_conn` is a list, not a class.

Reply:
But `self.waiting_conn` is a list of Worker which cannot directly apply the `wait` method.

Reply:
The ``wait`` interface is to wait a list of pending objects and to return a list of objects who are ready. This is async simulation, so we cannot let each individual connection to wait, we have to wait them all, and collect those ready ones.

Reply:
OK I see the issue. `wait` should be a method of of `VectorEnv` itself rather than the worker(s) from a design point of view. Howerver, iits implementation is worker-specific,a nd because of parallelism, it is not impossible to simply loop of each workers using a standard `for-loop`. I have to think about it, but I don't like the idea to ask to one worker to wait for its friends. It should be the pool itself to handle such high level behavior requiring awareness of the co-workers. I'm thinking how to handle it properly.

Reply:
> I don't like the idea to ask to one worker to wait for its friends

It is not one worker waiting for its friends. It is a static function of ``self.worker_class`` to wait a list of workers with the same type. Let the pool handle it is ok, but I'm concerned about the performance. See [my comment](https://github.com/thu-ml/tianshou/pull/179#discussion_r468045854)

Reply:
>  It is a static function of self.worker_class to wait a list of workers with the same type.

Yes I know, but it still feels like it is the workers that are handling it at the pool level. But I get your concern and indeed it must be avoided. Maybe using a Metaclass pattern could do the trick. I'm thinking.

Reply:
Any progress?

Reply:
I reply a few days ago, but it seems my comment was never published for some reason. Basically I said that it would be possible using something similar to template meta-programming in C++, but it would make the code significantly more complicated to understand and I think it is not a good idea.